python/Ganga/test/GPI/FileTests/TestMassStorage.py

```diff
@@ -0,0 +1,304 @@
+            assert os.path.isdir(output_dir)
+            for _input_file in j.inputfiles:
+                assert os.path.isfile(os.path.join(output_dir, _input_file.namePattern))
+
+        TestMassStorage.cleanUp()
+
+    def test_e_MultipleFiles(self):
+        """Test that the wildcards work"""
+
+        from Ganga.GPI import LocalFile, MassStorageFile, Job, ArgSplitter
+
+        _ext = '.root'
+        _ext2 = '.txt'
+        file_1 = generateUniqueTempFile(_ext)
+        file_2 = generateUniqueTempFile(_ext)
+        file_3 = generateUniqueTempFile(_ext2)
+        
+        j = Job()
+        j.inputfiles = [LocalFile(file_1), LocalFile(file_2), LocalFile(file_3)]
+        j.splitter = ArgSplitter(args = [[_] for _ in range(0, TestMassStorage.sj_len) ])
+        j.outputfiles = [MassStorageFile(namePattern='*'+_ext, outputfilenameformat='{jid}/{sjid}/{fname}'),
+                         MassStorageFile(namePattern='*'+_ext2)]
+        j.submit()
+
+    def test_f_MultiUpload(self):
+        """Test that multiple 'uploads' work"""
+
+        from Ganga.GPI import jobs
+        from Ganga.GPIDev.Base.Proxy import stripProxy
+
+        from GangaTest.Framework.utils import sleep_until_completed
+
+        j = jobs[-1]
+
+        sleep_until_completed(j)
+
+        assert j.status == 'completed'
+
+        assert len(j.subjobs) == TestMassStorage.sj_len
+
+        for i in range(0, TestMassStorage.sj_len):
+            # FIXME?
+            #assert (stripProxy(j.subjobs[i].outputfiles[0]).subfiles) == 2
+            output_dir = os.path.join(TestMassStorage.outputFilePath, str(j.id), str(i))
+            assert os.path.isdir(output_dir)
+            for file_ in j.inputfiles: 
+                assert os.path.isfile(os.path.join(output_dir, file_.namePattern))
+
+        TestMassStorage.cleanUp()
+
+    def test_g_MultipleFiles(self):
+        """Test that the wildcards work"""
+
+        from Ganga.GPI import LocalFile, MassStorageFile, Job, ArgSplitter
+
+        _ext = '.root'
+        file_1 = generateUniqueTempFile(_ext)
+        file_2 = generateUniqueTempFile(_ext)
+
+        j = Job()
+        j.inputfiles = [LocalFile(file_1), LocalFile(file_2)]
+        j.splitter = ArgSplitter(args = [[_] for _ in range(0, TestMassStorage.sj_len) ])
+        j.outputfiles = [MassStorageFile(namePattern='*'+_ext, outputfilenameformat='{jid}_{sjid}_{fname}')]
+        j.submit()
+
+    def test_h_MultiUpload(self):
+        """Test that multiple 'uploads' work"""
+
+        from Ganga.GPI import jobs
+        from Ganga.GPIDev.Base.Proxy import stripProxy
+
+        from GangaTest.Framework.utils import sleep_until_completed
+
+        j = jobs[-1]
+
+        sleep_until_completed(j)
+
+        assert j.status == 'completed'
+
+        assert len(j.subjobs) == TestMassStorage.sj_len
+
+        for i in range(0, TestMassStorage.sj_len):
+            # FIXME?
+            #assert (stripProxy(j.subjobs[i].outputfiles[0]).subfiles) == 2
```

Comment:
@drmarkwslater I'm aware this is how it's intended to be used and works but it does it by returning a lot of objects which until recently didn't correctly inherit the Job as a parent and still don't really do it in a full proof way because fixing this at the object level causes issues for some reason.
I can't see how to ever access the raw outputfiles to test this which is potentially an issue but one I'll leave for another day to worry about.